tests/test_components_pulses.py

```diff
@@ -449,6 +450,53 @@ def test_optical_laser_specials(mock_spb_aux_run):
     OpticalLaserPulses(run, ppl_seed=PPL_BITS.LP_SPB)
 
 
+@pytest.mark.parametrize('source', **pattern_sources)
+def test_machine_pulses_default(mock_spb_aux_run, source):
+    ref_pulses = MachinePulses(mock_spb_aux_run, source=source)
+    np.testing.assert_equal(ref_pulses.pulse_ids().iloc[:1350], np.r_[:2700:2])
+
+    pulse_counts = ref_pulses.pulse_counts()
+    assert not pulse_counts[:10].any()
+    assert (pulse_counts[10:] == 1350).all()
+
+
+def test_machine_pulses_specials(mock_spb_aux_run):
+    timeserver_run = mock_spb_aux_run.select('*TSYS/TIMESERVER*')
+
+    # Make an odd combination of bits, here SASE2 or soft kick.
+    odd_pulses = MachinePulses(timeserver_run,
+                               mask=DESTINATION_T5D | PHOTON_LINE_DEFLECTION)
+    pids = odd_pulses.pulse_ids()
+    assert pids.iloc[0] == 200  # The one SASE3 pulse.
+    np.testing.assert_equal(pids.iloc[1:64], np.r_[1500:2000:8])  # SASE2.
+
+    # Main dump and SPB laser seed.
+    odd_pulses = MachinePulses(timeserver_run, mode='and',
+                               mask=DESTINATION_TLD | PPL_BITS.LP_SPB)
+    np.testing.assert_equal(odd_pulses.pulse_ids().iloc[:50], np.r_[:300:6])
+
+    # Main dump and SA2.
+    odd_pulses = MachinePulses(timeserver_run, mode='and',
+                               mask=DESTINATION_TLD | DESTINATION_T5D)
+    assert odd_pulses.pulse_ids().empty
+
+    # Other types than string.
+    assert (MachinePulses(timeserver_run, mask=str(PHOTON_LINE_DEFLECTION)) \
```

Comment:
Why are strings allowed as a mask? :thinking: (it's not documented BTW)

Reply:
Oops, the comment meant to read `# Other types than int.`

The point was to test casting the mask to `int` in `MachinePulses.__init__`, which on the implementation side was simply to ensure no weird `TypeError`s. But it's not really intended behaviour to pass `str` as such, only the exception is relevant. Shall I remove that test?

Reply:
I think it's better to explicitly fail if anything other than an int is passed, if that's really the only intended argument type. If we allow strings then people might get confused since these are bit masks and try passing binary strings like `"001001"` without realizing those will be treated as decimal numbers.

Reply:
I remember now why I added the cast, I was afraid of `numpy`'s integer types, which can sometimes do weird things:

```python
In [5]: np.uint64(5) & (1 << 2)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[5], line 1
----> 1 np.uint64(5) & (1 << 2)

TypeError: ufunc 'bitwise_and' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''
``` 

I agree your point with strings of non-decimal integers... it would be weird to try, but indeed not produce the expected result. Shall I check against `isinstance(mask, np.integer)` before casting to `int` and raise a `TypeError` in all other cases?

Reply:
Yep, that sounds good :+1: